docs(aio): image sweep

petebacondarwin opened the pull request:
All the images in the guides have been checked and tidied up.
Fixes #16396

@wardbell : The following should become part of the "Authors Styleguide":

## Using Images

It is best to use HTML to declare images in the docs. Do not use the markdown `![...](...)`
shorthand.
The HTML to use is an `<img src="..." alt="...">` tag. You must supply a `src` attribute that is relative to the base path; and you should provide an `alt` attribute describing the image for accessibility. **Note that Image tags do not have a closing tag**.

### Image Size

The doc generation process will read the image dimensions and automatically add `width` and
`height` attributes to the `img` tag. If you want to control the size of the image then you should
supply your own `width` and `height` images.

Images should not be wider than 700px otherwise they may overflow the width of the document
in certain viewports. If you wish to have a larger image then provide a link to the actual image that
the user can click on to see the larger images separately.

### Image Formatting

There are three types of images that you can put in docs: inline, floating and standalone.

#### Inline Images

To create an inline image simply place the `img` tag in the content where you want the image to appear. E.g.

```markdown
The image here <img src="..." alt="..."> is visible inline in the text.
```

#### Floating images

You can cause an image to float to the left or right of the text by applying the `class="left"` or `class="right"` attributes respectively. E.g.

```markdown
<img src="..." alt="..." class="left">
This text will wrap around the to the right of this floating image.
```

All headings and code-examples will automatically clear a floating image.
If you need to force a piece of text to clear a floating image then you can use the following snippet:

```markdown
<br class="clear">
```

Finally, if you have floating images inside alerts or sub-sections then it is a good idea to apply the `clear-fix` class to the `div` to ensure that the image doesn't overflow its container. E.g.

```markdown
<div class="l-sub-section clear-fix">

  <img class="right" src="..." alt="...">
  Some **markdown** formatted text.

</div>
```

#### Standalone images

Some images should stand alone from the text. You do this by wrapping the `img` tag in a `figure`
tag. This causes the image to get additional styling, such as a rounded border and shadow. The text will not flow around this image but will stop before the image and start again afterword. E.g.

```markdown
Some paragraph preceding the image.

<figure>
  <img src="" alt="">
</figure>
```

(In the future we may also support styling for adding a caption to a standalone image.)

petebacondarwin committed:
fix(aio): allow code blocks to clear floated images

Previously the negative margin on the code headings were causing
floated images to overlay the start of a code block. Now all code block
successfully clear all floated elements.

mary-poppins commented:
The angular.io preview for c13cd257a1a88be3d5f58fe99a2d319763bea671 is available [here][1].

[1]: https://pr16609-c13cd257a1a88be3d5f58fe99a2d319763bea671.ngbuilds.io/

petebacondarwin committed:
feat(aio): add a `.clear` class for clearing floating images

mary-poppins commented:
The angular.io preview for 8294ad311c5579d3ab52a7e0c4cba6312a9d0276 is available [here][1].

[1]: https://pr16609-8294ad311c5579d3ab52a7e0c4cba6312a9d0276.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 025c3299b3ce3f7632f2983dc849a7c08fb2774f is available [here][1].

[1]: https://pr16609-025c3299b3ce3f7632f2983dc849a7c08fb2774f.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 9475ec7558e86ea946df54f9229ecd32efcb911f is available [here][1].

[1]: https://pr16609-9475ec7558e86ea946df54f9229ecd32efcb911f.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 383ad5e1c617989fbdeb6c111ccaa0b32cab5711 is available [here][1].

[1]: https://pr16609-383ad5e1c617989fbdeb6c111ccaa0b32cab5711.ngbuilds.io/

mary-poppins commented:
The angular.io preview for ff25d222bcc6437ab5f3a91dd85fcda48b2856c8 is available [here][1].

[1]: https://pr16609-ff25d222bcc6437ab5f3a91dd85fcda48b2856c8.ngbuilds.io/

petebacondarwin committed:
fix(aio): tidy up image styles

The css rules for `img.right` and `img.left` allow authors easy
access to floating an image on the left or right, respectively.

The `.image-display` rule which was always found on a figure
has been simplified so that all figures have this styling. It is very
unlikely that a figure will be used outside the content area; and
at this time it seems like `figure` is as good an indicator that we
want this kind of styling as anything.

Now that images are all tagged with width and height values, we cannot
assume to modify these dimensions via CSS as it can cause the image to
lose its correct proportions.  Until we find a better solition we must set
`height` to `auto` when the screen width is below 1300px to ensure that
these images maintain their proportions as they get shrunk to fit.

mary-poppins commented:
The angular.io preview for f487edb53a91e0b0f47aa3b84b2e2d100b0f58a0 is available [here][1].

[1]: https://pr16609-f487edb53a91e0b0f47aa3b84b2e2d100b0f58a0.ngbuilds.io/

petebacondarwin commented on aio/src/styles/2-modules/_images.scss:
```diff
@@ -1,25 +1,43 @@
-    float: none !important;
-  }
-}
+.content {
```
We only want this formatting of images within the content of documents, not the aio-shell in general.

petebacondarwin commented on aio/src/styles/2-modules/_images.scss:
```diff
@@ -1,25 +1,43 @@
+    @media (max-width: 1300px) {
+      max-width: 100%;
+      height: auto;
+    }
```
If the screen is wider than 1300px then we want to use the `width` and `height` values provided in the HTML to ensure that the page is laid out correctly while the images are still loading.

petebacondarwin replied:
But then as the screen narrows the images will be constrained by the `max-width` rule, which means that we need to switch to using `height: auto` to ensure that the proportions of the image do not change.

mary-poppins commented:
The angular.io preview for 151d7958471fb5b66dc89cb050537f90a49e0e16 is available [here][1].

[1]: https://pr16609-151d7958471fb5b66dc89cb050537f90a49e0e16.ngbuilds.io/

petebacondarwin committed:
docs(aio): use correct anchor for "back to the top" link

The `#toc` anchor does not work when the page is
wide enough that the TOC is floating to the side.

mary-poppins commented:
The angular.io preview for 68263a12d33bdd7f94b73a08ea9b258d44d526de is available [here][1].

[1]: https://pr16609-68263a12d33bdd7f94b73a08ea9b258d44d526de.ngbuilds.io/

petebacondarwin committed:
build(aio): add `#top-of-page` to path variants for link checking

Since the `#top-of-page` is outside the rendered docs
the `checkAnchorLinks` processor doesn't find them
as valid targets for links.
Adding them as a `pathVariant` solves this problem
but will still catch links to docs that do not actually exist.

gkalpak commented on aio/content/guide/architecture.md:
```diff
@@ -452,57 +322,38 @@ as with event binding.
 
 <figure>
-  <img src="generated/images/guide/architecture/parent-child-binding.png" alt="Parent/Child binding" style="float:left; width:300px; margin-left:-40px;margin-right:10px"></img>
+  <img src="generated/images/guide/architecture/parent-child-binding.png" alt="Parent/Child binding" width="400">
```
Nit: I think the idea was to make this image the same size as the previous one (which looks better). This behavior is not retained here.

petebacondarwin replied:
agreed - will fix

petebacondarwin replied:
![screen shot 2017-05-09 at 23 07 35](https://cloud.githubusercontent.com/assets/15655/25874835/53506676-350c-11e7-8d66-80fcac948bd0.png)

gkalpak commented on aio/tools/transforms/angular.io-package/index.js:
```diff
@@ -42,4 +42,5 @@ module.exports = new Package('angular.io', [gitPackage, apiPackage, contentPacka
         return (existsSync(resolve(SRC_PATH, url)));
       }
     });
+    checkAnchorLinksProcessor.pathVariants = ['', '/', '.html', '/index.html', '#top-of-page'];
```
What is `.html` for?

petebacondarwin replied:
The `.html` is only there because the defaults for that processor were: `['', '/', '.html', '/index.html']`, so I am really just adding one new one. :-)

petebacondarwin replied:
you'd have never known if I'd just used `.push('#....')`.

gkalpak replied:
Don't be so sure :wink:

gkalpak reviewed (approved):
LGTM (with a few minor comments/questions).

mary-poppins commented:
The angular.io preview for f66dc2eda2fdfcbbd444e3a973efd2bc79921a0d is available [here][1].

[1]: https://pr16609-f66dc2eda2fdfcbbd444e3a973efd2bc79921a0d.ngbuilds.io/

wardbell reviewed (commented):
Same question. I do like baking #top-of-page though. 

Otherwise, lgtm

Foxandxss commented on aio/content/guide/reactive-forms.md:
```diff
@@ -459,8 +459,8 @@ The `heroForm.value` returns the _form model_.
-<figure class='image-display'>
-  <img src="generated/images/guide/reactive-forms/json-output.png" width="400px" alt="JSON output"></img>
+<figure>
+  <img src="generated/images/guide/reactive-forms/json-output.png" width="400" height="133" alt="JSON output">
```
I was going to comment this one too.

petebacondarwin replied:
For each image I set the width to the suggested width and then checked that the browser resized the height to. I may have got it wrong in this case. It should have been `176`.

Foxandxss commented on aio/content/guide/set-document-title.md:
```diff
@@ -29,7 +29,7 @@ See the <live-example name="set-document-title"></live-example>.
 
     <td>
-      <img src='generated/images/plunker/plunker-switch-to-editor-button.png' width="200px" height="70px" alt="pop out the window" align="right"></img>      <br></br>      <img src='generated/images/plunker/plunker-separate-window-button.png' width="200px" height="47px" alt="pop out the window" align="right"></img>
+      <img src='generated/images/plunker/plunker-switch-to-editor-button.png' width="200" height="70" alt="pop out the window" align="right">      <br></br>      <img src='generated/images/plunker/plunker-separate-window-button.png' width="200" height="47" alt="pop out the window" align="right">
```
This in narrow mode is almost invisible.

petebacondarwin replied:
Hmm, that is because the image is in a `td` and it's width gets set to `max-width: 100%` but that it 100% of the `td` rather than the page. We might need some additional styling to cope with this situation.

petebacondarwin replied:
Actually using a table is not right here. It is much better to use a subsection and floating images. I'll fix up.

Foxandxss commented:
The embedded plunker preview images are suuuuuuuuuper big.

petebacondarwin commented on aio/content/guide/aot-compiler.md:
```diff
@@ -605,7 +605,7 @@ showing exactly which application and Angular modules and classes are included i
-  <figure class='image-display'>
-    <img src="generated/images/guide/aot-compiler/toh-pt6-bundle.png" alt="toh-pt6-bundle"></img>
+  <figure>
+    <img src="generated/images/guide/aot-compiler/toh-pt6-bundle.png" alt="toh-pt6-bundle" width="700">
```
This raw image is super big. A width of 700px is about all you can manage on a screen that is at the breakpoint of 1300px before the `max-width: 100%` kicks in.

gkalpak replied:
Shouldn't we resize the image then (and save the cost of downloading it)?

petebacondarwin replied:
Yes but that was outside of my remit :-P
I'll create an issue to track it.

petebacondarwin replied:
Seee #16678

petebacondarwin commented on aio/content/guide/upgrade.md:
```diff
@@ -339,8 +339,8 @@ everything work seamlessly:
-<figure class='image-display'>
-  <img src="generated/images/guide/upgrade/injectors.png" alt="The two injectors in a hybrid application" width="700"></img>
+<figure>
+  <img src="generated/images/guide/upgrade/injectors.png" alt="The two injectors in a hybrid application" width="700" height="262">
```
The actual image is too large for the width of the page at the breakpoint where we switch to `max-width: 100%` and `height: auto`. 700 is the biggest width that works at this screen width.

petebacondarwin commented on aio/src/styles/2-modules/_images.scss:
```diff
@@ -1,25 +1,43 @@
+
+    &.right {
+      float: right;
+      margin-left: 20px;
```
Good call, we need to lose the floating when narrow, I guess.

petebacondarwin replied:
![screen shot 2017-05-09 at 23 08 16](https://cloud.githubusercontent.com/assets/15655/25874856/699b2808-350c-11e7-82c5-5970358f6d24.png)

petebacondarwin committed:
fix(aio): ensure that headings clear floated images

petebacondarwin committed:
fix(aio): do not force live-example embedded image to 100% size

This made them look too big, generally. Leaving them with no size means
that they will look reasonable in large viewports and switch to 100% width
in narrow viewports.

petebacondarwin commented:
I fixed the embedded plunker images.

petebacondarwin committed:
docs(aio): general tidy up of image HTML in guides

Previously, the guides have a lot of inline image styling and unnecessary
use of the `image-display` css class.
Images over 700px are problematic for guide docs, so those have been given
specific widths and associated heights.

petebacondarwin commented:
I think this is all fixed up @Foxandxss and @gkalpak PTAL

mary-poppins commented:
The angular.io preview for 84b16a19f03049bd82acce22cb2937a59aa2b892 is available [here][1].

[1]: https://pr16609-84b16a19f03049bd82acce22cb2937a59aa2b892.ngbuilds.io/

mary-poppins commented:
The angular.io preview for 1802c5ce9c4ca898c83274c371caa8841f7fa3ba is available [here][1].

[1]: https://pr16609-1802c5ce9c4ca898c83274c371caa8841f7fa3ba.ngbuilds.io/

Foxandxss commented:
Works for me pete.
